fdas/main.go

```diff
@@ -0,0 +1,321 @@
+		}
+		var newACL fdasACL
+		if err := c.BindJSON(&newACL); err != nil {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid ACL definition"})
+			return
+		}
+		parts := strings.Split(newACL.Object, ":")
+		if len(parts) != 2 {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid ACL object definition"})
+			return
+		}
+		namespace := parts[0]
+		pair, _, err := kv.Get(namespace, nil)
+		if err != nil || pair == nil {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid ACL namespace"})
+			return
+		}
+		availableRelations := strings.Split(string(pair.Value), ",")
+		if !checkIfRelationValid(newACL.Relation, &availableRelations) {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "ACL relation not supported in this namespace"})
+			return
+		}
+
+		err = db.Put([]byte(newACL.User+"/"+newACL.Object), []byte(newACL.Relation), nil)
+		if err != nil {
+			c.IndentedJSON(http.StatusInternalServerError, gin.H{"message": "internal server error"})
+			return
+		}
+
+		c.IndentedJSON(http.StatusCreated, newACL)
+
+	}
+}
+
+func checkACL(tokens *Set, db *leveldb.DB, kv *capi.KV) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if !tokens.Has(c.Param("token")) {
+			c.IndentedJSON(http.StatusUnauthorized, gin.H{"message": "unauthorized"})
+			return
+		}
+		object := c.Query("object")
+		relation := c.Query("relation")
+		user := c.Query("user")
+
+		if object == "" || relation == "" || user == "" {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid ACL request"})
+			return
+		}
+
+		data, err := db.Get([]byte(user+"/"+object), nil)
+		if err != nil {
+			c.IndentedJSON(http.StatusNotFound, gin.H{"message": "ACL not found"})
+			return
+		}
+		if string(data) == relation {
+			c.IndentedJSON(http.StatusOK, true)
+			return
+		}
+		parts := strings.Split(object, ":")
+		if len(parts) != 2 {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid ACL object definition"})
+			return
+		}
+		namespace := parts[0]
+		value, _, err := kv.Get(namespace+"/"+string(data)+"/"+relation, nil)
+		if err == nil || value != nil {
+			c.IndentedJSON(http.StatusOK, true)
+		} else {
+			c.IndentedJSON(http.StatusOK, false)
+		}
+
+	}
+}
+
+func postNamespace(tokens *Set, kv *capi.KV) gin.HandlerFunc {
+	return func(c *gin.Context) {
+		if !tokens.Has(c.Param("token")) {
+			c.IndentedJSON(http.StatusUnauthorized, gin.H{"message": "unauthorized"})
+			return
+		}
+		var newNamespace fdasConfig
+		relatshionshipMap := make(map[string][]string)
+		allRelations := ""
+		if err := c.BindJSON(&newNamespace); err != nil {
+			c.IndentedJSON(http.StatusBadRequest, gin.H{"message": "invalid namespace definition"})
+			return
+		}
+		pair, _, err := kv.Get(newNamespace.Namespace, nil)
+		if err == nil || pair != nil {
+			keys, _, err := kv.Keys(newNamespace.Namespace, "", nil)
+			if err != nil {
+				c.IndentedJSON(http.StatusInternalServerError, gin.H{"message": "internal server error"})
```

Comment:
Non-graceful fall

Reply:
Noted will be changed in next commit. Thank you